examples/custom_storage.ml

```diff
@@ -0,0 +1,114 @@
+(*
+ * Copyright (c) 2022 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Lwt.Syntax
+
+module Hashtbl_config = struct
+  module Conf = Irmin.Backend.Conf
+
+  let spec = Conf.Spec.v "hashtbl"
+  let init_size = Conf.key ~spec "init-size" Irmin.Type.int 8
+  let empty = Conf.empty spec
+end
+
+module Hashtbl_storage : Irmin.Simple_storage.Maker =
+functor
+  (Key : Irmin.Type.S)
+  (Value : Irmin.Type.S)
+  ->
+  struct
+    module Tbl = Hashtbl.Make (struct
+      type t = Key.t
+
+      let equal a b = Irmin.Type.(unstage (equal Key.t)) a b
+      let hash k = Irmin.Type.(unstage (short_hash Key.t)) k
+    end)
+
+    (** Types *)
+
+    type 'a t = Value.t Tbl.t
+    type key = Key.t
+    type value = Value.t
+
+    (** Initialisation / Closing *)
+
+    let v config =
+      let init_size = Irmin.Backend.Conf.get config Hashtbl_config.init_size in
+      Tbl.create init_size |> Lwt.return
+
+    let close _t = Lwt.return_unit
+
+    (** Operations *)
+
+    let set t key value = Tbl.replace t key value |> Lwt.return
+    let mem t key = Tbl.mem t key |> Lwt.return
+    let find t key = Tbl.find_opt t key |> Lwt.return
+    let keys t = Tbl.to_seq_keys t |> List.of_seq |> Lwt.return
+    let remove t key = Tbl.remove t key |> Lwt.return
+    let clear t = Tbl.clear t |> Lwt.return
+    let cast t = (t :> Irmin.Perms.read_write t)
+    let lock = Mutex.create ()
```

Comment:
I like that idea. I tried doing that previously but ran into type checking issues (related to the cast) that I couldn't resolve. Thankfully, this will resolve by moving the lock to `t`.